ui: reintroduce end-to-end UI tests with cypress

sjbarag opened the pull request:
A recent commit [1] removed the stale, unused end-to-end tests. In their
place, add newly-written tests that can run headlessly in TeamCity (via
Docker) as a quick-failing health check and as a more exhaustive suite.

[1] 3d171b234b (ui: remove unused end-to-end UI tests, 2022-07-25)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85095)
<!-- Reviewable:end -->

sjbarag commented:
do-not-merge added because I need to merge https://github.com/cockroachdb/yarn-vendored/pull/98 first :)

sjbarag commented:
Don't get scared by the line count here — most of that is a new `yarn.lock`!

sjbarag commented:
I guess I should add a README.md to that tree huh 😅

laurenbarker reviewed (approved):
I have some questions, but overall this <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Someone else should do a pass on the docker/bazel stuff who has relevant knowledge.

> Reviewed 25 of 26 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-:-N84kcMXCE3jyuBotphn:bngncn2)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sjbarag and @srosenberg)

___
*[`build/teamcity/cockroach/ci/tests/ui_e2e_test_impl.sh` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84fgdO9rJS5SDjtS1-:-N84fgdO9rJS5SDjtS10:b1diehh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/build/teamcity/cockroach/ci/tests/ui_e2e_test_impl.sh#L3)):*
> ```Shell
> #!/usr/bin/env bash
> function build_docker_image() {
>   # Buffer noisy output and only print it on failure.
> ```

Smart! Do we do this in other places?
___
*[`build/teamcity/cockroach/ci/tests/ui_e2e_test_impl.sh` line 15 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84frin9PvN_fgaa9Y-:-N84frin9PvN_fgaa9Y0:b-jbcc6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/build/teamcity/cockroach/ci/tests/ui_e2e_test_impl.sh#L15)):*
> ```Shell
>   SPEC_ARG=""
>   if [ "health" = "${$1:='EMPTY'}" ]; then
>     SPEC_ARG="--spec 'cypress/e2e/health-check/**'"
> ```

Is there a reason to specify the directory? Does `--e2e` not work because the tests are nested in the `health-check` directory?
___
*[`pkg/ui/workspaces/e2e-tests/.eslintrc.json` line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84gPsG8FNzIQgGWWx-:-N84gPsG8FNzIQgGWWx0:by4t2hf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/.eslintrc.json#L13)):*
> ```jsonc
>     "@typescript-eslint/no-inferrable-types": "off",
>     "@typescript-eslint/ban-types": "off",
>     "@typescript-eslint/no-unused-vars": "off",
> ```

Why are these rules overridden to off? Wouldn't we want these on?
___
*[`pkg/ui/workspaces/e2e-tests/Dockerfile` line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84gzxH2uZYp3VgY9xL:-N84gzxH2uZYp3VgY9xM:b-qwgt7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/Dockerfile#L28)):*
> ```Dockerfile
> #     cockroachdb/cockroach-ci-ui
> 
> FROM cypress/browsers:node16.14.2-slim-chrome100-ff99-edge
> ```

Ah, maybe we should use this image in managed-service too, so the cypress version in CI always matches what we use locally :thinking: 
___
*[`pkg/ui/workspaces/e2e-tests/Dockerfile` line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84hJf1DPBIPHlFcaaE:-N84hJf1DPBIPHlFcaaF:b-eqhcul) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/Dockerfile#L43)):*
> ```Dockerfile
> 
> # Use the compiled CRDB binary from a previous TC job via Artifact Dependency:
> # TODO(barag): include a TC link here
> ```

Is this a TODO for post-merge?
___
*[`pkg/ui/workspaces/e2e-tests/prettier.config.js` line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84hoUPD8_C_RSDsTN3:-N84hoUPD8_C_RSDsTN4:bc01nt4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/prettier.config.js#L12)):*
> ```JavaScript
> 
> module.exports = {
>   trailingComma: "all",
> ```

Our shared eslint config includes prettier rules. Do we need this config?
___
*[`pkg/ui/workspaces/e2e-tests/cypress/fixtures/example.json` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84iS1c-tfNBeeGXhA-:-N84iS1c-tfNBeeGXhA0:b-rppa5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/fixtures/example.json#L2)):*
> ```JSON
> {
>   "name": "Using fixtures to represent data",
> ```

Are these being used? It looks like dummy data.
___
*[`pkg/ui/workspaces/e2e-tests/cypress/plugins/index.ts` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84iw6y3oQdSruMOd_J:-N84iw6y3oQdSruMOd_K:bguyfy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/plugins/index.ts#L29)):*
>The plugins file is no longer supported as of Cypress version 10.0.0.
> https://docs.cypress.io/guides/references/migration-guide#Plugins-File-Removed

Was this file added by default? It's no longer supported :thinking: 
___
*[`pkg/ui/workspaces/e2e-tests/cypress/support/commands.ts` line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84jM2yApgpWedbvrrs:-N84jM2yApgpWedbvrrt:b-ej41i7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/support/commands.ts#L33)):*
> ```XML
> });
> 
> import "@testing-library/cypress/add-commands";
> ```

Why is this at the bottom?
___
*[`pkg/ui/workspaces/e2e-tests/cypress/support/index.ts` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84jo5T3cmyp_mhL3Qb:-N84jo5T3cmyp_mhL3Qc:b-z45vtg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/support/index.ts#L41)):*
> ```XML
> }
> 
> import "./commands";
> ```

Should the imports be at the bottom?


<!-- Sent from Reviewable.io -->

sjbarag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85095)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @laurenbarker, @sjbarag, and @srosenberg)

___
*[`build/teamcity/cockroach/ci/tests/ui_e2e_test_impl.sh` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84fgdO9rJS5SDjtS1-:-N85O6oU5hJ-qeO572ro:bssiwwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/build/teamcity/cockroach/ci/tests/ui_e2e_test_impl.sh#L3)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Smart! Do we do this in other places?
</blockquote></details>

Thanks, but I get no credit!  I copied this from a few other files in `build/teamcity/cockroach/…` :smile: 
___
*[`build/teamcity/cockroach/ci/tests/ui_e2e_test_impl.sh` line 15 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84frin9PvN_fgaa9Y-:-N85OB59-_ILEyy4q41R:b-co0u40) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/build/teamcity/cockroach/ci/tests/ui_e2e_test_impl.sh#L15)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Is there a reason to specify the directory? Does `--e2e` not work because the tests are nested in the `health-check` directory?
</blockquote></details>

By default `--e2e` runs every test it finds!  That works perfectly, but there's a use-case where we'll want to run only the health-check tests within that `health-check` directory, as part of a short-running job per-PR.
___
*[`pkg/ui/workspaces/e2e-tests/.eslintrc.json` line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84gPsG8FNzIQgGWWx-:-N85OmEABxKBIavi1KXw:bed19bi) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/.eslintrc.json#L13)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Why are these rules overridden to off? Wouldn't we want these on?
</blockquote></details>

TBH I copied them from `db-console` :) I'll see if I can just take the defaults without much trouble (there's not much code here, so it'll probably be easy)
___
*[`pkg/ui/workspaces/e2e-tests/Dockerfile` line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84hJf1DPBIPHlFcaaE:-N85Orzw8Bji_nFSh0jn:bb77uf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/Dockerfile#L43)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Is this a TODO for post-merge?
</blockquote></details>

Yes!
___
*[`pkg/ui/workspaces/e2e-tests/prettier.config.js` line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84hoUPD8_C_RSDsTN3:-N85Osm1DJ5PSaPwsUzR:b-2qs2g1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/prettier.config.js#L12)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Our shared eslint config includes prettier rules. Do we need this config?
</blockquote></details>

Probably not!  I'll remove it and let you know.
___
*[`pkg/ui/workspaces/e2e-tests/cypress/fixtures/example.json` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84iS1c-tfNBeeGXhA-:-N85OuvVDueK9XO0H2Qa:b-5rhcq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/fixtures/example.json#L2)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Are these being used? It looks like dummy data.
</blockquote></details>

Correct, this was generated by default.  I can remove it.
___
*[`pkg/ui/workspaces/e2e-tests/cypress/plugins/index.ts` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84iw6y3oQdSruMOd_J:-N85Owhq1IfiTeclXHpW:b-uvsemp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/plugins/index.ts#L29)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

>The plugins file is no longer supported as of Cypress version 10.0.0.
> https://docs.cypress.io/guides/references/migration-guide#Plugins-File-Removed

Was this file added by default? It's no longer supported :thinking: 
</blockquote></details>

It was!  That's hilarious, this was a from-scratch integration.  I'll remove it.
___
*[`pkg/ui/workspaces/e2e-tests/cypress/support/commands.ts` line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84jM2yApgpWedbvrrs:-N85P-sV6aPaKV7hUHGY:b-3smivi) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/support/commands.ts#L33)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Why is this at the bottom?
</blockquote></details>

:shrug: that was default — I'll hoist it!
___
*[`pkg/ui/workspaces/e2e-tests/cypress/support/index.ts` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84jo5T3cmyp_mhL3Qb:-N85P1f5C4L_oxy6Kj8I:b-arulxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/support/index.ts#L41)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Should the imports be at the bottom?
</blockquote></details>

Probably not!  I'll hoist it.  This was just the default location for Cypress.


<!-- Sent from Reviewable.io -->

nathanstilwell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85095)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @laurenbarker, @sjbarag, and @srosenberg)

___
*[`pkg/ui/workspaces/e2e-tests/.eslintrc.json` line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84gPsG8FNzIQgGWWx-:-N85Y0po2Vrq9Vc1hmV2:b-cqu3e9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/.eslintrc.json#L13)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

TBH I copied them from `db-console` :) I'll see if I can just take the defaults without much trouble (there's not much code here, so it'll probably be easy)
</blockquote></details>

I would suggest we drop them, as we have these rules suppressed in DB Console for legacy code. Since we are starting over with the E2E tests we should hold them to a higher standard.


<!-- Sent from Reviewable.io -->

sjbarag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85095)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @laurenbarker, @sjbarag, and @srosenberg)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N85aRoN2kmaHe3eAlbf:-N85aRoN2kmaHe3eAlbg:badw80w) (no related file):*
Thanks for the review, @laurenbarker !

Changed in r2:
* a README.md for these tests
* support for `dev ui e2e --headed`, so that working with `./dev` is at parity with direct `yarn` invocations
* tests for `dev ui e2e`

___


<!-- Sent from Reviewable.io -->

sjbarag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85095)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @laurenbarker, @sjbarag, and @srosenberg)

___
*[`pkg/ui/workspaces/e2e-tests/.eslintrc.json` line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84gPsG8FNzIQgGWWx-:-N85aioY-ifaZlQEon4H:b157x7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/.eslintrc.json#L13)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

I would suggest we drop them, as we have these rules suppressed in DB Console for legacy code. Since we are starting over with the E2E tests we should hold them to a higher standard.
</blockquote></details>

Fixed in r3
___
*[`pkg/ui/workspaces/e2e-tests/cypress/support/commands.ts` line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84jM2yApgpWedbvrrs:-N85ateR0wVMeKhKn88u:b157x7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/support/commands.ts#L33)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

:shrug: that was default — I'll hoist it!
</blockquote></details>

Fixed in r3
___
*[`pkg/ui/workspaces/e2e-tests/cypress/support/index.ts` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84jo5T3cmyp_mhL3Qb:-N85b27X8e4Q5d7JOtlJ:b157x7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/support/index.ts#L41)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

Probably not!  I'll hoist it.  This was just the default location for Cypress.
</blockquote></details>

Fixed in r3
___
*[`pkg/ui/workspaces/e2e-tests/prettier.config.js` line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84hoUPD8_C_RSDsTN3:-N85azIf6TDOYCcmpBG2:b-toc2y1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/prettier.config.js#L12)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

Probably not!  I'll remove it and let you know.
</blockquote></details>

Removed in r3
___
*[`pkg/ui/workspaces/e2e-tests/cypress/fixtures/example.json` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84iS1c-tfNBeeGXhA-:-N85algm30BNR_zwd-og:b-toc2y0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/fixtures/example.json#L2)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

Correct, this was generated by default.  I can remove it.
</blockquote></details>

Removed in r4
___
*[`pkg/ui/workspaces/e2e-tests/cypress/plugins/index.ts` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N84iw6y3oQdSruMOd_J:-N85awPK0-S7Ec7_1tgh:b-toc2y1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db07af1bd68908aad4f72a12b78674c40a69f15/pkg/ui/workspaces/e2e-tests/cypress/plugins/index.ts#L29)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

It was!  That's hilarious, this was a from-scratch integration.  I'll remove it.
</blockquote></details>

Removed in r3


<!-- Sent from Reviewable.io -->

laurenbarker reviewed (approved):
> Reviewed 6 of 9 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85095)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @laurenbarker, @sjbarag, and @srosenberg)

___
*[`pkg/ui/workspaces/e2e-tests/README.md` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N85eNLi7fjTnp_XITwN:-N85eNLi7fjTnp_XITwO:bdbl8a0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9d06355bfb5d19b19f1dd91d66b64b9b769f09e/pkg/ui/workspaces/e2e-tests/README.md#L1)):*
> ```Markdown
> # End-to-end Tests for CockroachDB UI
> ```

Very nice docs!
___
*[`pkg/ui/workspaces/e2e-tests/README.md` line 53 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85095#-N85eDz4-1ODLaCrNVhn:-N85eDz4-1ODLaCrNVho:b-csemiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9d06355bfb5d19b19f1dd91d66b64b9b769f09e/pkg/ui/workspaces/e2e-tests/README.md#L53)):*
> ```Markdown
> ## Health Checks vs All Tests
> Each CockroachDB pull request results in only the tests in
> `./cypress/e2e/health-checks` being executed. These are intended to confirm only
> ```

Ah, now I get it


<!-- Sent from Reviewable.io -->

sjbarag commented:
As a demo, here's a passing health-check suite: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_ScratchProjectPutTcExperimentsInHere_WipBaragCypressHealth/5936309

Once this merges, we can make the necessary TeamCity changes to copy this configuration into the proper cockroach/ci/tests folder and add a nearly-identical one that runs `build/teamcity/cockroach/ci/tests/ui_e2e_test_all.sh`.

rickystewart commented on pkg/ui/workspaces/e2e-tests/Dockerfile:
```diff
@@ -0,0 +1,51 @@
+# WARNING:
+# This Dockerfile must be built from the root of cockroachdb.git with a prebuilt
+# cockroach binary at ./upstream_artifactsl/cockroach. Running it from other
```
`upstream_artifacts`

rickystewart commented on pkg/ui/workspaces/e2e-tests/Dockerfile:
```diff
@@ -0,0 +1,51 @@
+#     -v $PWD/artifacts:/artifacts \
+#     cockroachdb/cockroach-ci-ui
+
+FROM cypress/browsers:node16.14.2-slim-chrome100-ff99-edge
```
I'm pretty done with tests that build a Docker image every time they run. Building the image takes a long time and it's just a matter of time until something skews so things start failing. Can we pre-build this once and push it to our DockerHub?

sjbarag replied:
> Building the image takes a long time and it's just a matter of time until something skews so things start failing.

Well the good news is these are pretty static / minimal images.  The build here took 1m 28s in https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_ScratchProjectPutTcExperimentsInHere_WipBaragCypressHealth/5936309?buildTab=log&focusLine=3&linesState=232, but that's still longer than the tests themselves.  We could probably build and publish an image that's `cypress/browsers:node16.14.2-slim-chrome100-ff99-edge` + `apt install libresolv-wrapper` only, then inject a `cockroach` binary and the `pkg/ui/workspaces/e2e-tests` tree at runtime and  run `yarn install && yarn test` inside the container.

When we want to test against newer browser versions in CI, we'd have to rebuild and republish that image but that shouldn't be so bad.  I'll message you about the mechanics of getting that image published!

sjbarag replied:
I mean, we could also go full-pancake and not maintain our own image at all:

```
  docker run \
    --rm \
    -it \
    -e 'IS_DOCKER=1' \
    -e 'COCKROACH=/upstream_artifacts/cockroach' \
    -v $PWD:/upstream_artifacts \
    -v $PWD/pkg/ui/workspaces/e2e-tests:/cypress \
    cypress/browsers:node16.14.2-slim-chrome100-ff99-edge \
    /bin/bash -c 'apt install libresolv-wrapper; pushd /cypress; yarn install --force; yarn test'
```

runs locally, which really isn't so bad 🤔

srosenberg replied:
I agree with @rickystewart. We shouldn't be relying on fetching docker images and ideally all other types of artifacts from the public internet. There is always a chance of that artifact being temporarily unavailable either due to a rate limit (e.g., [1]) or a transient (public) network issue; this will inevitably lead to flaky tests. Not to mention the security aspect... always a chance of a BGP spoofing attack to serve you rogue containers.


[1] https://docs.docker.com/docker-hub/download-rate-limit/#what-is-the-download-rate-limit-on-docker-hub

srosenberg commented on pkg/ui/workspaces/e2e-tests/build/start-crdb-movr.sh:
```diff
@@ -0,0 +1,89 @@
+  --multitenant=false \
+  --http-port=8080 \
+  --listening-url-file=$CONN_URL_FILE \
+  --execute 'select pg_sleep(1000000)' &
```
Curious why `movr`? Isn't intended for interactive sessions? Why not `start-singlenode --insecure`?

sjbarag replied:
`start-single-node` required a significant amount of setup before the tests could start executing — mostly generating certs, creating a non-root user with a password, creating a few tables, etc.  `cockroach workload` helps with that slightly, but doing it manually winds up being quite a bit slower than `cockroach demo movr` as well.  I found an early attempt at that and it was slower and more effort for no benefit :) https://gist.github.com/sjbarag/30eb3c3e253235fef00c461affa1348f

Curiously, a lot of that is handled in [build/deploy/cockroach.sh](https://github.com/cockroachdb/cockroach/blob/e52f732f8130df956449786c053d1d4539548411/build/deploy/cockroach.sh) (the entrypoint for the `cockroachdb/cockroach` and `cockroachdb/cockroach-ci` docker images).  I've got a WIP implementation that uses `docker compose` for these purposes in CI, so I'll see if I can make that script work for dev uses as well.

srosenberg replied:
>Curiously, a lot of that is handled in [build/deploy/cockroach.sh](https://github.com/cockroachdb/cockroach/blob/e52f732f8130df956449786c053d1d4539548411/build/deploy/cockroach.sh)

Right, `democluster.NewDemoCluster` does essentially the same bootstrapping as in `cockroach.sh` but then switches into `ShellCtx`. It looks like this predates `cockroach.sh` [1] which was added relatively recently and expressly for use with Docker :)

[1] https://github.com/cockroachdb/cockroach/commit/2862374a4743aece5707ea52d0ac17f6cb10dc7b

> I've got a WIP implementation that uses docker compose for these purposes in CI, so I'll see if I can make that script work for dev uses as well.

I think that would be nicer, thanks!

srosenberg commented on pkg/ui/workspaces/e2e-tests/README.md:
```diff
@@ -0,0 +1,58 @@
+Cypress to allow testing of both authenticated and unauthenticated flows.
+
+## Health Checks vs All Tests
+Each CockroachDB pull request results in only the tests in
```
Nice! Is the idea to run these as a separate TC job as opposed to adding them to `UI Tests` which are mostly unit tests?

sjbarag replied:
Correct!  Here's a sample: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_ScratchProjectPutTcExperimentsInHere_WipBaragCypressHealth/5936309

sjbarag commented:
@rickystewart / @srosenberg how would you feel about merging this before TeamCity gets support for `docker compose` (via https://cockroachlabs.atlassian.net/browse/DEVINF-442)?  I'd like to at least get this out for devs to write tests for so it doesn't skew into oblivion 😃

srosenberg commented:
> @rickystewart / @srosenberg how would you feel about merging this before TeamCity gets support for `docker compose` (via https://cockroachlabs.atlassian.net/browse/DEVINF-442)? I'd like to at least get this out for devs to write tests for so it doesn't skew into oblivion 😃

I am in support of merging. The stability period starts on Monday; the extra safety net wrt UI e2e tests is going to give us more confidence.

sjbarag commented:
bors r=laurenbarker,nathanstilwell,rickystewart,srosenberg

sjbarag commented:
booooooo.  @rickystewart looks like we need your approval to unblock this merge ❤️

rickystewart commented on build/teamcity/cockroach/ci/tests/ui_e2e_test_impl.sh:
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
+function load_cockroach_docker_image() {
+  docker load --input upstream_artifacts/cockroach-docker-image.tar.gz &> artifacts/docker-load.log || (cat artifacts/docker-load.log && false)
```
Nifty.

rickystewart reviewed (commented):
Bumping the version in `./dev` seems like a good idea.

sjbarag commented:
Dev bumped to 53.  Thanks y'all!

sjbarag commented:
bors r=laurenbarker,nathanstilwell,rickystewart,srosenberg

sjbarag commented:
bors r+

rickystewart commented:
@sjbarag The CLA check didn't work, you might try doing a no-op commit --amend and re-pushing to re-trigger it.

sjbarag committed:
ui: reintroduce end-to-end UI tests with cypress

A recent commit [1] removed the stale, unused end-to-end tests. In their
place, add newly-written tests that can run headlessly in TeamCity (via
Docker) as a quick-failing health check and as a more exhaustive suite.

[1] 3d171b234b (ui: remove unused end-to-end UI tests, 2022-07-25)

Release note: None

sjbarag commented:
> @sjbarag The CLA check didn't work, you might try doing a no-op commit --amend and re-pushing to re-trigger it.

Strange!  Rebased on master, so let's give this another go:

bors r=laurenbarker,nathanstilwell,rickystewart,srosenberg

(🤔 do I need to specify reviewers again, or is `r+` enough? I can never remember)
